### PR TITLE
Gh-380: Update user guide querying section

### LINKED ISSUE #380: Update Querying section in the user guide
Following the restructure work on #378 a new section in the user guide has been created for querying. This section needs updating with the relevant information on the available gaffer APIs, much of this information could be taken from existing sections  (e.g. current getting started guide) but some new bits may need to be added.

### PR DESCRIPTION
**Pull request by tb06904**
Quite a few changes on this one broken out the pages on the APIs so they each have their own page as an introduction and how to get up and running. Fleshed out the 'querying' section of the user guide with examples and information on the Gaffer querying syntax, including operations and filtering etc.

Its possible more information could be added here in future (e.g. gremlin) but this should serve as a good basis and cover most of what a user might need to know to get familiar with queries.

Note some white space alterations due to merge conflict resolving.

# Related Issue

- Resolve #380

### CODE REVIEW

File: docs/user-guide/apis/java-api.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+to utilise this form of the API.
+
+## Querying a Graph
+
+Using Java to Query a graph unlike the other APIs requires a reference to a
+`Graph` object that essentially represents a graph.
+
+With the other APIs you would connect directly to a running instance via the
+rest interface; however, to do this with Java you would need to configure a
+`Graph` object with a proxy store.
```
Comment (GCHQDeveloper314) [inline docs/user-guide/apis/java-api.md]: We could add a link to the reference page for the Proxy Store.

File: docs/user-guide/apis/rest-api.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+commonly by the provided [Swagger UI](https://swagger.io/).
+
+## Querying a Graph
+
+If you wish to just query to get some information about the graph instance such
+as, what schema it is using or what available Operations it has then, there
+should already be `GET` endpoints to do that. Executing any of these `GET`
+requests will simply 'get' you some information but wont allow any input.
+
+To query and extract custom bits of data from the graph there is an endpoint at
```
Comment (GCHQDeveloper314) [inline docs/user-guide/apis/rest-api.md]: Not sure about this wording. The main way users will query the graph is with the execute endpoint. Other endpoints will be of limited interest. It feels like this could be made more prominent.

File: docs/user-guide/query/gaffer-syntax/import-export/csv.md [hunk 3]
```diff
@@ -1,43 +1,68 @@
-# CSV Import and Export
+# CSV Data Import and Export
 
-!!! info "Work in Progress"
+Gaffer supports both importing from and exporting to csv. This page will outline
```
Comment (GCHQDeveloper314) [inline docs/user-guide/query/gaffer-syntax/import-export/csv.md]: To me, "will outline" seems to make it sound incomplete, compared to "outlines".

File: docs/user-guide/apis/rest-api.md (hunk 2 above)
Comment (tb06904) [reply docs/user-guide/apis/rest-api.md]: Fair point this page is more a general background on the API but I could reword to make the execute endpoint more the focus

Comment (GCHQDeveloper314) [reply docs/user-guide/apis/rest-api.md]: As this is in the user guide I think that would make sense. Users don't need anything other than the execute endpoint. They could use the schema endpoint, but that can also be done with an operation. I don't think it needs more content, just a change in focus to the execute endpoint.

File: docs/user-guide/apis/java-api.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# Using the Java API
+
+As Gaffer is written in Java there is native support to allow use of all its
+public classes. Using Gaffer via the Java interface does differ from the rest
+API and `gafferpy` but is fully featured with extensive
+[Javadocs](https://gchq.github.io/Gaffer/overview-summary.html). However, You
```
Comment (cn337131) [inline docs/user-guide/apis/java-api.md]:
```suggestion
[Javadocs](https://gchq.github.io/Gaffer/overview-summary.html). However, you
```

File: docs/user-guide/apis/java-api.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+will of course need to be familiar with writing and running Java code in order
+to utilise this form of the API.
+
+## Querying a Graph
+
+Using Java to Query a graph unlike the other APIs requires a reference to a
```
Comment (cn337131) [inline docs/user-guide/apis/java-api.md]: not sure if query needs to be capitalised?

File: docs/user-guide/apis/python-api.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# Using the Python API
+
+This section covers an overview of the python API extension for Gaffer to
+demonstrate how to get up and running to perform queries from Python code on an
+existing running graph.
+
+!!! tip
+    Please see the handy introduction on [what is python](../gaffer-basics/what-is-python.md)
```
Comment (cn337131) [inline docs/user-guide/apis/python-api.md]:
```suggestion
    Please see the handy introduction to [Python](../gaffer-basics/what-is-python.md)
```
Potential rephrase?

File: docs/user-guide/apis/python-api.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+Commonly referred to as `gafferpy` this is an API to gaffer that provides
+similar querying capabilities to the rest API but from Python. Fundamentally it
+wraps the rest API to use the same JSON under the hood this means you should be
+able to access almost any features or end points available in the main rest API.
+
+## Installation
+
+Currently there isn't a release of `gafferpy` on pypi or other pip repository;
+however, the source code can still be cloned from the [git repository](https://github.com/gchq/gaffer-tools/tree/develop/python-shell)
+and installed via pip. Please see the readme in the `gafferpy` repository for
```
Comment (cn337131) [inline docs/user-guide/apis/python-api.md:22]: link the readme?

File: docs/user-guide/apis/python-api.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+g_connector = gaffer_connector.GafferConnector("http://localhost:8080/rest/latest")
+```
+
+Once connected you can access and run the same endpoints and operations as you
+would via the usual rest API but via their python classes. The endpoints are
+accessed via the `GafferConnector` to allow you executing Operation chains to
+perform queries on the graph.
+
+!!! note
+    There may be of the features of the full rest API not be present in
```
Comment (cn337131) [inline docs/user-guide/apis/python-api.md]:
```suggestion
    Some of the features of the full rest API may not be present in
```

File: docs/user-guide/apis/python-api.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    `gafferpy` as always check the [reference guide](../../reference/intro.md)
```
Comment (cn337131) [inline docs/user-guide/apis/python-api.md]:
```suggestion
    `gafferpy` so always check the [reference guide](../../reference/intro.md)
```

File: docs/user-guide/apis/python-api.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    an example, the following will get all the elements in a graph:
+
+    ```python
+    elements = g_connector.execute_operation(
+        operation =  gaffer.GetAllElements()
+    )
+    ```
+
+!!! example ""
+    An Operation chain can be ran using the `execute_operation_chain()` function.
```
Comment (cn337131) [inline docs/user-guide/apis/python-api.md]:
```suggestion
    An Operation chain can be run using the `execute_operation_chain()` function.
```

File: docs/user-guide/apis/python-api.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            operations=[
+                gaffer.GetAllElements(),
+                gaffer.Count()
+            ]
+        )
+    )
+    ```
+
+Along with running Operations you can of course optionally apply some filtering
+provided by a `View`. There are more detail on both [filtering](../query/gaffer-syntax/filtering.md)
```
Comment (cn337131) [inline docs/user-guide/apis/python-api.md]:
```suggestion
using a `View`. There is more detail on both [filtering](../query/gaffer-syntax/filtering.md)
```

Not sure if I understand it right, but provided by a View does not seem to make sense. Might be a rephrase on the whole sentence instead.

File: docs/user-guide/apis/rest-api.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
+# Using the Rest API
+
+These sections will cover the usage of the Gaffer rest API to perform queries
+and operations on a graph. This guide should cover a lot of the use cases a user
+may face; however, please refer to the [reference guide](../../reference/intro.md)
```
Comment (cn337131) [inline docs/user-guide/apis/rest-api.md]:
```suggestion
may face; however please refer to the [reference guide](../../reference/intro.md)
```

File: docs/user-guide/apis/rest-api.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+In Gaffer, the Rest API consists of various predefined HTTP requests known as
+endpoints that can be used to interact with a running graph instance. These
+endpoints are accessed either by sending a crafted HTTP request to them e.g.
+with a tool like [`curl`](https://curl.se/docs/httpscripting.html) or more
+commonly by the provided [Swagger UI](https://swagger.io/).
+
+## Querying a Graph
+
+If you wish to just query to get some information about the graph instance such
+as, what schema it is using or what available Operations it has then, there
```
Comment (cn337131) [inline docs/user-guide/apis/rest-api.md]:
```suggestion
as what schema it is using or what available Operations it has then there
```

File: docs/user-guide/apis/python-api.md (hunk 7 above)
Comment (tb06904) [reply docs/user-guide/apis/python-api.md:22]: I could do, but I didn't put it in here as there was a link to the repository in the sentence directly before which will render the readme when you access it

File: docs/user-guide/apis/python-api.md (hunk 11 above)
Comment (tb06904) [reply docs/user-guide/apis/python-api.md]: Was probably meant to read as "apply some filtering by providing a `View`"

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 14]
```diff
@@ -0,0 +1,520 @@
+# Filtering Data
+
+One of the key parts of querying data is being able to filter bits out you may
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
One of the key parts of querying data is being able to filter out bits you may
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+not want. This page will cover some of the common ways you can filter the
+returned data from an Operation to get the data you want.
+
+In Gaffer the main way you filter data is by applying whats known as a `View` to
+a returned set of elements. A `View` as the name suggests, allows you to view
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
a returned set of elements. A `View`, as the name suggests, allows you to view
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+the data in a different way this can be either via a filter, aggregation,
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
the data in a different way; this can be either via a filter, aggregation,
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+    -->
+    C(["Software
+
+        ID: 2"])
+```
+
+Now say the `weight` property could represent how much a `Person` contributed to
+creating something, and so we wanted to find only significant contributions. To
+do this we can apply a filter to act as a threshold to only get edges with a
+`weight` more than a specific value.
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]: The whole paragraph doesn't read too well. I would rephrase it to flow a little better. 
```suggestion

Lets say the `weight` property represents how much a `Person` contributed to
creating something and that we wanted to only find the most significant contributions. To
do this we can apply a filter to act as a threshold to only get edges with a
`weight` more than a specific value.
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+            .input(new EntitySeed("John"))
+            .view(viewWithFilters)
+            .build();
+
+        graph.execute(operation, user);
+        ```
+
+To form relevant filters and queries its almost required that you are aware of
+the graphs schema that is in use as this determines what properties and elements
+you can reference in your queries. For an introduction and background on
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]: This sentence is a little confusing and could do with clarifying

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 19]
```diff
@@ -0,0 +1,520 @@
...
+    else statements in traditional programming. For a full list of available
+    predicates refer to the [reference documentation](../../../reference/predicates-guide/predicates.md).
+
+### Filtering Properties
+
+If you are only interested in specific properties then it is more efficient to
+tell Gaffer to only return those properties. This can be easily achieved by
+applying the `properties` or `excludeProperties` field to a `View`.
+
+If we take a similar example as before but instead add a couple of properties
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
If we take a similar example to before but instead add a couple of properties
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 20]
```diff
@@ -0,0 +1,520 @@
...
+     weight: 0.6
+     hours: 800"
+    -->
+    C(["Software
+
+        ID: 2"])
+```
+
+Now as before we can run a query on this graph to get the elements relevant to
+the `Person` node however, this time we will filter so that only specific
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
the `Person` node. However, this time we will filter so that only specific
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 21]
```diff
@@ -0,0 +1,520 @@
...
+            .input(new EntitySeed("John"))
+            .view(viewWithFilters)
+            .build();
+
+        graph.execute(operation, user);
+        ```
+
+## Transformation
+
+Its possible to apply a transformation to the output of a query this gives you
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
It is possible to apply a transformation to the output of a query which then gives you the opportunity
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 22]
```diff
@@ -0,0 +1,520 @@
...
+an opportunity to manipulate the results into a more useful output.
+
+When a transform is applied the new results are save into whats known as a
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
When a transform is applied the new results are saved into what is known as a
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 23]
```diff
@@ -0,0 +1,520 @@
...
+transient property. A transient property is just a property that is not
+persisted, simply created at query time by a transform function.
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
persisted; simply created at query time by a transform function.
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 24]
```diff
@@ -0,0 +1,520 @@
...
+                            .transformer(new MultiplyBy(60))
+                            .build())
+                    .build())
+            .build();
+
+        graph.execute(getEdgesWithMinutes, user);
+        ```
+
+The `selection` in a transform is similar to the way we select properties and
+identifiers in a filter, and as demonstrated you can select (and also project)
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
identifiers in a filter and as demonstrated you can select (and also project)
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 25]
```diff
@@ -0,0 +1,520 @@
...
+any property but also or any of these unique identifiers:
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
any property but also any of these unique identifiers:
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 26]
```diff
@@ -0,0 +1,520 @@
...
+
+- `VERTEX` - This is the vertex on an Entity.
+- `SOURCE` - This is the source vertex on an Edge.
+- `DESTINATION` - This is the destination vertex on an Edge.
+- `DIRECTED` - This is the directed field on an Edge.
+- `MATCHED_VERTEX` - This is the vertex that was matched in the query, either
+  the `SOURCE` or the `DESTINATION`.
+- `ADJACENT_MATCHED_VERTEX` - This is the adjacent vertex that was matched in
+  the query, either the `SOURCE` or the `DESTINATION`. For example, if your seed
+  matches the source of the edge This would resolve to the `DESTINATION` value.
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
  matches the source of the edge this would resolve to the `DESTINATION` value.
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 27]
```diff
@@ -0,0 +1,520 @@
...
+
+## Query-time Aggregation
+
+Gaffer allows grouping results together based on their properties to form a new
+result, this is known as aggregation. Aggregation can be applied at both data
+ingest or query-time, this guide will focus on the latter but an overview of
+both techniques is available in the [gaffer basics
+guide](../../gaffer-basics/what-is-aggregation.md).
+
+Generally to apply aggregation at query-time you must override whats known as
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
Generally to apply aggregation at query-time you must override what is known as
```

File: docs/user-guide/query/gaffer-syntax/operations.md [hunk 28]
```diff
@@ -0,0 +1,157 @@
+# Operations
+
+Querying in Gaffer is all mainly based around whats called "Operations". You can
+think of these as lots of small functions/commands that can be either standalone
+or chained together into an Operation Chain to form more complex query
```
Comment (cn337131) [inline docs/user-guide/query/gaffer-syntax/operations.md]:
```suggestion
or be chained together into an Operation Chain to form more complex query
```

**Review by cn337131 — APPROVED** [review]

File: docs/user-guide/query/gaffer-syntax/operations.md [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+like the following:
+
+```json
+{
+    "class": "Operation",
+    "input": "Input"
+}
+```
+
+As you can see its fairly simple, we first define the Operation we wish to use
```
Comment (GCHQDeveloper314) [inline docs/user-guide/query/gaffer-syntax/operations.md]: "it's", though I think it could just be:
```suggestion
Using Operations is fairly simple, we first define the Operation we wish to use
```

File: docs/user-guide/query/gaffer-syntax/operations.md [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+
+!!! tip
+    To get a list of all available Operations for a graph you can use the
+    endpoint `/graph/operations`
+
+## Operation Chains
+
+The Operations in Gaffer can be chained together to form complex queries on the
+graph. This page will give some general example usage of how you can chain
+Operations together.
```
Comment (GCHQDeveloper314) [inline docs/user-guide/query/gaffer-syntax/operations.md]:
```suggestion
The Operations in Gaffer can be chained together to form complex graph queries.
This page will give some general example usage of how you can chain Operations
together.
```

File: docs/user-guide/query/gaffer-syntax/operations.md [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+        ```java
+        OperationChain<Long> countElements = new OperationChain.Builder()
+            .first(new GetElements.Builder().input(new EntitySeed("v1")).build())
+            .then(new Count<>())
+            .build();
+
+        Long result = graph.execute(countElements, user);
+        ```
+
+To chain operations its important to take note of what each operations input and
```
Comment (GCHQDeveloper314) [inline docs/user-guide/query/gaffer-syntax/operations.md]:
```suggestion
To chain operations it's important to take note of what each operations input and
```

File: docs/user-guide/apis/rest-api.md [hunk 32]
```diff
@@ -0,0 +1,45 @@
+# Using the Rest API
+
+These sections will cover the usage of the Gaffer rest API to perform queries
+and operations on a graph. This guide should cover a lot of the use cases a user
+may face; however please refer to the [reference guide](../../reference/intro.md)
+for a full list of whats possible.
```
Comment (GCHQDeveloper314) [inline docs/user-guide/apis/rest-api.md]:
```suggestion
for a full list of what's possible.
```
Preferably:
```suggestion
for a full list of what is possible.
```

File: docs/user-guide/query/gaffer-syntax/filtering.md [hunk 33]
```diff
@@ -0,0 +1,520 @@
...
+persisted; simply created at query time by a transform function.
+
+To use a transform you must use a transform function, this is a Java class
+that extends the [`java.util.Function`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html)
+class to take some input and give a new output. Commonly you would want to
+write your own transform function class as it can be quite specific to your
+graph data and what analytics you want to get out of it. However, there is
+the [Koryphe module](../../../reference/functions-guide/koryphe-functions.md)
+that is included by default with Gaffer which provides a some default functions
+you can make use of.
```
Comment (GCHQDeveloper314) [inline docs/user-guide/query/gaffer-syntax/filtering.md]:
```suggestion
graph data and what analytics you want to get out of it. However, the
[Koryphe module](../../../reference/functions-guide/koryphe-functions.md)
(included by default with Gaffer) provides some default functions you can
make use of.
```

File: docs/user-guide/apis/java-api.md [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+`Graph` object that essentially represents a graph.
+
+With the other APIs you would connect directly to a running instance via the
+rest interface; however, to do this with Java you would need to configure a
+`Graph` object with a [proxy store](../../administration-guide/gaffer-stores/proxy-store.md).
+
+!!! example ""
+    The following example uses the `ProxyStore.Builder()` to configure a `Graph`
+    to connect to the required address (in this case
+    `http://localhost:8080/rest/latest`).
```
Comment (GCHQDeveloper314) [inline docs/user-guide/apis/java-api.md]: Why the choice of `/rest/latest`? I think `/rest/` is the usual endpoint, or at least the ones used by the docker images?

File: docs/user-guide/apis/java-api.md [hunk 35]
```diff
@@ -0,0 +1,59 @@
...
+
+If you're interested in learning more about the `Graph` class please see the
+[developers guide](../../development-guide/project-structure/components/graph.md)
+on the subject.
+
+!!! note
+    One other thing to note about the Java API is that the `execute()` method
+    requires you to have a `User`. This is part of the available fine grain
+    security Gaffer provides see the [admin guide](../../administration-guide/security/security-guide.md)
+    for more information.
```
Comment (GCHQDeveloper314) [inline docs/user-guide/apis/java-api.md]:
```suggestion
    requires you to have a `User`. This is part of Gaffer's fine grain security,
    see the [admin guide](../../administration-guide/security/security-guide.md)
    for more information.
```

**Review by GCHQDeveloper314 — COMMENTED** [review]
Few minor comments. The filtering guide is good.

File: docs/user-guide/apis/java-api.md (hunk 34 above)
Comment (tb06904) [reply docs/user-guide/apis/java-api.md]: Think this might've been a typo from the v1 docs I'll change

**Review by GCHQDeveloper314 — APPROVED** [review]
